File: model/index-vocab.py
import os
import re
from collections import defaultdict

def index_corpus(corpus_dir, output_file):
    # Initialize vocabulary and posting lists
    vocabulary = set()
    posting_lists = defaultdict(list)

    # Regular expression pattern to match special characters
    special_characters_pattern = r'[!?.,$;\'-[\]><&{}@\(\)\"]'

    # Iterate through each document in the corpus directory
    for filename in os.listdir(corpus_dir):
        with open(os.path.join(corpus_dir, filename), 'r', encoding='latin-1') as file:
            content = file.read().lower().split()  # Read content and convert to lowercase
            doc_id = filename.split('.')[0]  # Extract document ID from filename

            # Update vocabulary and posting lists
            for token in content:
                # Remove special characters from the token
                token = re.sub(special_characters_pattern, '', token)
                if token:
                    vocabulary.add(token)
                    posting_lists[token].append(doc_id)

    # Sort vocabulary alphabetically
    sorted_vocabulary = sorted(vocabulary)

    # Write posting lists to the output file
    with open(output_file, 'w') as outfile:
        for token in sorted_vocabulary:
            posting = ' '.join(posting_lists[token])
            outfile.write(f"{token}\t{posting}\n")

def search(index, term):
    if term in index:
        return index[term]
    else:
        return []

def search_and(index, term1, term2):
    posting1 = set(search(index, term1))
    posting2 = set(search(index, term2))
    result = posting1.intersection(posting2)
    return sorted(list(result))

def search_or(index, term1, term2):
    posting1 = set(search(index, term1))
    posting2 = set(search(index, term2))
    result = posting1.union(posting2)
    return sorted(list(result))

def search_and_not(index, term1, term2):
    posting1 = set(search(index, term1))
    posting2 = set(search(index, term2))
    result = posting1.difference(posting2)
    return sorted(list(result))

def search_or_not(index, term1, term2):
    posting1 = set(search(index, term1))
    posting2 = set(search(index, term2))
    result = posting1.symmetric_difference(posting2)
    return sorted(list(result))


if __name__ == "__main__":
    # Define input and output paths
    dir = "/home/phanh/Downloads/study/ir" # Change this with your own directory
    corpus_directory = f"{dir}/reuters/test"
    index_file = f"{dir}/index.txt"

    # Index the corpus and generate the posting lists
    index_corpus(corpus_directory, index_file)
    print("Indexing completed. Posting lists saved to index.txt")

    index = {}
    with open(index_file, 'r') as file:
        for line in file:
            term, postings = line.strip().split('\t')
            index[term] = postings.split()

    # Perform searches based on different query types
    print("Single-term Query:")
    print("america:", search(index, 'america'))

    print("\nTwo-term Query with AND:")
    print("america AND oil:", search_and(index, 'america', 'oil'))

    print("\nTwo-term Query with OR:")
    print("america OR oil:", search_or(index, 'america', 'oil'))

    print("\nTwo-term Query with AND and NOT:")
    print("america AND (NOT oil):", search_and_not(index, 'america', 'oil'))

    print("\nTwo-term Query with OR and NOT:")
    print("america OR (NOT oil):", search_or_not(index, 'america', 'oil'))

File: index-vocab.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class Indexing {

    public static void indexCorpus(String corpusDirectory, String indexFile) throws IOException {
        Set<String> uniqueWords = new HashSet<>();
        Map<String, List<String>> postingLists = new HashMap<>();
        String specialCharactersPattern = "[!?.,;'\\[\\]{}@\\(\\)\"]";

        File[] files = new File(corpusDirectory).listFiles();
        if (files != null) {
            for (File file : files) {
                String fileName = file.getName();
                try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
                    String line;
                    StringBuilder contentBuilder = new StringBuilder();
                    while ((line = reader.readLine()) != null) {
                        contentBuilder.append(line.toLowerCase()).append(" ");
                    }
                    String content = contentBuilder.toString();
                    String documentId = fileName.split("\\.")[0];
                    for (String token : content.split("\\s+")) {
                        token = token.replaceAll(specialCharactersPattern, "");
                        if (!token.isEmpty()) {
                            uniqueWords.add(token);
                            postingLists.computeIfAbsent(token, k -> new ArrayList<>()).add(documentId);
                        }
                    }
                }
            }
        }

        List<String> sortedUniqueWords = new ArrayList<>(uniqueWords);
        Collections.sort(sortedUniqueWords);

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(indexFile))) {
            for (String token : sortedUniqueWords) {
                String posting = String.join(" ", postingLists.get(token));
                writer.write(token + "\t" + posting + "\n");
            }
        }
    }

    public static List<String> search(Map<String, List<String>> index, String queryTerm) {
        return index.getOrDefault(queryTerm, new ArrayList<>());
    }

    public static List<String> searchAnd(Map<String, List<String>> index, String term1, String term2) {
        Set<String> posting1 = new HashSet<>(search(index, term1));
        Set<String> posting2 = new HashSet<>(search(index, term2));
        posting1.retainAll(posting2);
        return new ArrayList<>(posting1);
    }

    public static List<String> searchOr(Map<String, List<String>> index, String term1, String term2) {
        Set<String> posting1 = new HashSet<>(search(index, term1));
        Set<String> posting2 = new HashSet<>(search(index, term2));
        posting1.addAll(posting2);
        return new ArrayList<>(posting1);
    }

    public static List<String> searchAndNot(Map<String, List<String>> index, String term1, String term2) {
        Set<String> posting1 = new HashSet<>(search(index, term1));
        Set<String> posting2 = new HashSet<>(search(index, term2));
        posting1.removeAll(posting2);
        return new ArrayList<>(posting1);
    }

    public static List<String> searchOrNot(Map<String, List<String>> index, String term1, String term2) {
        Set<String> posting1 = new HashSet<>(search(index, term1));
        Set<String> posting2 = new HashSet<>(search(index, term2));
        posting1.addAll(posting2);
        posting1.removeAll(searchAnd(index, term1, term2));
        return new ArrayList<>(posting1);
    }

    public static void main(String[] args) throws IOException {
        String directory = "D:/data_mining"; // Thay đổi đường dẫn này thành thư mục chứa tập văn bản của bạn.
        String corpusDirectory = directory + "/reuters/test";
        String indexFile = directory + "/index.txt";

        // Index the corpus and generate the posting lists
        indexCorpus(corpusDirectory, indexFile);
        System.out.println("Indexing completed. Posting lists saved to index.txt");

        // Các dòng code sau đây không cần thay đổi.

    Map<String, List<String>> index = new HashMap<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(indexFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split("\t");
                index.put(parts[0], Arrays.asList(parts[1].split("\\s+")));
            }
        }

        // Perform searches based on different query types
        System.out.println("Single-term Query:");
        System.out.println("america: " + search(index, "america"));

        System.out.println("\nTwo-term Query with AND:");
        System.out.println("america AND oil: " + searchAnd(index, "america", "oil"));

        System.out.println("\nTwo-term Query with OR:");
        System.out.println("america OR oil: " + searchOr(index, "america", "oil"));

        System.out.println("\nTwo-term Query with AND and NOT:");
        System.out.println("america AND (NOT oil): " + searchAndNot(index, "america", "oil"));

        System.out.println("\nTwo-term Query with OR and NOT:");
        System.out.println("america OR (NOT oil): " + searchOrNot(index, "america", "oil"));
    }
}
